File: sdks/sandbox/go/endpoint_cache.go
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package opensandbox

import (
	"container/list"
	"context"
	"fmt"
	"sync"
	"time"

	"golang.org/x/sync/singleflight"
)

const (
	DefaultEndpointCacheTTL  = 600 * time.Second
	DefaultEndpointCacheSize = 1024
)

type endpointCacheKey struct {
	sandboxID      string
	port           int
	useServerProxy bool
}

func (k endpointCacheKey) String() string {
	return fmt.Sprintf("%s:%d:%v", k.sandboxID, k.port, k.useServerProxy)
}

type endpointCacheEntry struct {
	key       endpointCacheKey
	endpoint  *Endpoint
	expiresAt time.Time
}

func cloneEndpoint(ep *Endpoint) *Endpoint {
	headers := make(map[string]string, len(ep.Headers))
	for k, v := range ep.Headers {
		headers[k] = v
	}
	return &Endpoint{Endpoint: ep.Endpoint, Headers: headers, Origin: ep.Origin}
}

// EndpointCache is a thread-safe LRU+TTL cache for sandbox endpoints.
type EndpointCache struct {
	mu      sync.Mutex
	entries map[endpointCacheKey]*list.Element
	order   *list.List
	maxSize int
	ttl     time.Duration

	group singleflight.Group
}

// NewEndpointCache creates a cache with the given max size and TTL.
// If maxSize <= 0, DefaultEndpointCacheSize is used.
// If ttl <= 0, DefaultEndpointCacheTTL is used.
func NewEndpointCache(maxSize int, ttl time.Duration) *EndpointCache {
	if maxSize <= 0 {
		maxSize = DefaultEndpointCacheSize
	}
	if ttl <= 0 {
		ttl = DefaultEndpointCacheTTL
	}
	return &EndpointCache{
		entries: make(map[endpointCacheKey]*list.Element),
		order:   list.New(),
		maxSize: maxSize,
		ttl:     ttl,
	}
}

// Get retrieves a cached endpoint. Returns nil, false on miss or expiry.
func (c *EndpointCache) Get(key endpointCacheKey) (*Endpoint, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	elem, ok := c.entries[key]
	if !ok {
		return nil, false
	}
	entry := elem.Value.(*endpointCacheEntry)
	if time.Now().After(entry.expiresAt) {
		c.removeLocked(elem)
		return nil, false
	}
	c.order.MoveToFront(elem)
	return cloneEndpoint(entry.endpoint), true
}

// Put stores an endpoint in the cache, evicting LRU entries if at capacity.
func (c *EndpointCache) Put(key endpointCacheKey, ep *Endpoint) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if elem, ok := c.entries[key]; ok {
		c.order.MoveToFront(elem)
		entry := elem.Value.(*endpointCacheEntry)
		entry.endpoint = cloneEndpoint(ep)
		entry.expiresAt = time.Now().Add(c.ttl)
		return
	}

	for c.order.Len() >= c.maxSize {
		oldest := c.order.Back()
		if oldest == nil {
			break
		}
		c.removeLocked(oldest)
	}

	entry := &endpointCacheEntry{
		key:       key,
		endpoint:  cloneEndpoint(ep),
		expiresAt: time.Now().Add(c.ttl),
	}
	elem := c.order.PushFront(entry)
	c.entries[key] = elem
}

// Invalidate removes all cached and inflight entries for the given sandbox ID.
func (c *EndpointCache) Invalidate(sandboxID string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	for key, elem := range c.entries {
		if key.sandboxID == sandboxID {
			c.removeLocked(elem)
			c.group.Forget(key.String())
		}
	}
}

// Len returns the number of entries in the cache.
func (c *EndpointCache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.order.Len()
}

func (c *EndpointCache) removeLocked(elem *list.Element) {
	entry := elem.Value.(*endpointCacheEntry)
	delete(c.entries, entry.key)
	c.order.Remove(elem)
}

// GetOrFetch checks cache, deduplicates inflight requests via singleflight,
// and calls fetch on miss. The caller's context is respected: if ctx is
// cancelled while waiting for a shared inflight request, the caller returns
// immediately with the context error (the shared fetch continues for other waiters).
func (c *EndpointCache) GetOrFetch(ctx context.Context, key endpointCacheKey, fetch func() (*Endpoint, error)) (*Endpoint, error) {
	if ep, ok := c.Get(key); ok {
		return ep, nil
	}

	ch := c.group.DoChan(key.String(), func() (interface{}, error) {
		if ep, ok := c.Get(key); ok {
			return ep, nil
		}
		ep, err := fetch()
		if err != nil {
			return nil, err
		}
		c.Put(key, ep)
		return ep, nil
	})

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case result := <-ch:
		if result.Err != nil {
			return nil, result.Err
		}
		return cloneEndpoint(result.Val.(*Endpoint)), nil
	}
}

File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/infrastructure/cache/EndpointCache.kt
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.alibaba.opensandbox.sandbox.infrastructure.cache

import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxEndpoint
import java.time.Duration
import java.util.LinkedHashMap
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

data class EndpointCacheKey(
    val sandboxId: String,
    val port: Int,
    val useServerProxy: Boolean,
)

private data class CacheEntry(
    val endpoint: SandboxEndpoint,
    val expiresAtNanos: Long,
)

/**
 * Thread-safe LRU + TTL endpoint cache with inflight deduplication.
 */
class EndpointCache(
    maxSize: Int = 1024,
    private val ttl: Duration = Duration.ofSeconds(600),
) {
    private val maxSize = maxOf(1, maxSize)
    private val lock = ReentrantLock()
    private var generation = 0L

    private val cache =
        object : LinkedHashMap<EndpointCacheKey, CacheEntry>(16, 0.75f, true) {
            override fun removeEldestEntry(eldest: MutableMap.MutableEntry<EndpointCacheKey, CacheEntry>?): Boolean {
                return size > this@EndpointCache.maxSize
            }
        }

    private val inflight = mutableMapOf<EndpointCacheKey, InflightEntry>()

    private fun cloneEndpoint(ep: SandboxEndpoint) =
        SandboxEndpoint(
            endpoint = ep.endpoint,
            headers = ep.headers.toMap(),
            origin = ep.origin,
        )

    private class InflightEntry {
        @Volatile var result: SandboxEndpoint? = null

        @Volatile var error: Exception? = null
        val latch = java.util.concurrent.CountDownLatch(1)
    }

    fun get(key: EndpointCacheKey): SandboxEndpoint? =
        lock.withLock {
            val entry = cache[key] ?: return null
            if (System.nanoTime() >= entry.expiresAtNanos) {
                cache.remove(key)
                return null
            }
            cloneEndpoint(entry.endpoint)
        }

    fun put(
        key: EndpointCacheKey,
        endpoint: SandboxEndpoint,
    ) = lock.withLock {
        cache[key] = CacheEntry(cloneEndpoint(endpoint), System.nanoTime() + ttl.toNanos())
    }

    fun invalidate(sandboxId: String) =
        lock.withLock {
            generation++
            cache.keys.removeAll { it.sandboxId == sandboxId }
            inflight.keys.removeAll { it.sandboxId == sandboxId }
        }

    fun getOrFetch(
        key: EndpointCacheKey,
        fetch: () -> SandboxEndpoint,
    ): SandboxEndpoint {
        get(key)?.let { return it }

        val existingInflight: InflightEntry?
        val myInflight: InflightEntry?
        val genBefore: Long

        lock.withLock {
            // Double-check
            val entry = cache[key]
            if (entry != null && System.nanoTime() < entry.expiresAtNanos) {
                return cloneEndpoint(entry.endpoint)
            }
            if (entry != null) cache.remove(key)

            val existing = inflight[key]
            if (existing != null) {
                existingInflight = existing
                myInflight = null
                genBefore = -1
            } else {
                val inf = InflightEntry()
                inflight[key] = inf
                existingInflight = null
                myInflight = inf
                genBefore = generation
            }
        }

        if (existingInflight != null) {
            existingInflight.latch.await()
            existingInflight.error?.let { throw it }
            return existingInflight.result!!
        }

        try {
            val endpoint = fetch()
            lock.withLock {
                if (generation == genBefore) {
                    cache[key] = CacheEntry(cloneEndpoint(endpoint), System.nanoTime() + ttl.toNanos())
                }
            }
            myInflight!!.result = endpoint
            return endpoint
        } catch (e: Exception) {
            myInflight!!.error = e
            throw e
        } finally {
            lock.withLock { inflight.remove(key) }
            myInflight!!.latch.countDown()
        }
    }

    val size: Int get() = lock.withLock { cache.size }
}
